File: server/src/models/donation.model.js
import mongoose from 'mongoose';

const { Schema } = mongoose;

const donationSchema = new Schema(
  {
    donorId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    donorName: {
      type: String,
      required: true,
    },
    bloodType: {
      type: String,
      required: true,
      match: /^(A|B|AB|O)[+-]$/,
      uppercase: true,
    },
    requestId: {
      type: Schema.Types.ObjectId,
      ref: 'BloodRequest',
    },
    facilityId: {
      type: Schema.Types.ObjectId,
      ref: 'Facility',
      required: true,
    },
    facilityName: {
      type: String,
      required: true,
    },
    quantity: {
      type: Number,
      default: 1,
      min: 1,
    },
    donationDate: {
      type: Date,
      default: Date.now,
    },
    status: {
      type: String,
      enum: ['scheduled', 'completed', 'cancelled'],
      default: 'scheduled',
    },
    notes: String,
  },
  { timestamps: true },
);

donationSchema.index({ donorId: 1 });
donationSchema.index({ requestId: 1 });
donationSchema.index({ facilityId: 1 });
donationSchema.index({ donationDate: -1 });
donationSchema.index({ status: 1 });

export default mongoose.model('Donation', donationSchema);

File: lib/models/donation_model.dart
/// Model representing a blood donation record
class DonationModel {
  final String id;
  final String donorId;
  final String donorName;
  final String bloodType;
  final String facilityId;
  final String facilityName;
  final int units;
  final DateTime donationDate;
  final String status; // 'completed', 'pending', 'cancelled'
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  DonationModel({
    required this.id,
    required this.donorId,
    required this.donorName,
    required this.bloodType,
    required this.facilityId,
    required this.facilityName,
    required this.units,
    required this.donationDate,
    required this.status,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory DonationModel.fromJson(Map<String, dynamic> json) {
    return DonationModel(
      id: json['_id'] ?? json['id'] ?? '',
      donorId: json['donorId'] ?? json['donor_id'] ?? '',
      donorName: json['donorName'] ?? json['donor_name'] ?? '',
      bloodType: json['bloodType'] ?? json['blood_type'] ?? '',
      facilityId: json['facilityId'] ?? json['facility_id'] ?? '',
      facilityName: json['facilityName'] ?? json['facility_name'] ?? '',
      units: json['units'] ?? 1,
      donationDate: json['donationDate'] != null
          ? DateTime.parse(json['donationDate'])
          : json['donation_date'] != null
              ? DateTime.parse(json['donation_date'])
              : DateTime.now(),
      status: json['status'] ?? 'pending',
      notes: json['notes'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : json['created_at'] != null
              ? DateTime.parse(json['created_at'])
              : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : json['updated_at'] != null
              ? DateTime.parse(json['updated_at'])
              : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'donorId': donorId,
      'donorName': donorName,
      'bloodType': bloodType,
      'facilityId': facilityId,
      'facilityName': facilityName,
      'units': units,
      'donationDate': donationDate.toIso8601String(),
      'status': status,
      'notes': notes,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  DonationModel copyWith({
    String? id,
    String? donorId,
    String? donorName,
    String? bloodType,
    String? facilityId,
    String? facilityName,
    int? units,
    DateTime? donationDate,
    String? status,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return DonationModel(
      id: id ?? this.id,
      donorId: donorId ?? this.donorId,
      donorName: donorName ?? this.donorName,
      bloodType: bloodType ?? this.bloodType,
      facilityId: facilityId ?? this.facilityId,
      facilityName: facilityName ?? this.facilityName,
      units: units ?? this.units,
      donationDate: donationDate ?? this.donationDate,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
